streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -342,118 +353,36 @@ public void close(final boolean clean) throws ProcessorStateException {
     }
 
     @Override
-    public void checkpoint(final Map<TopicPartition, Long> checkpointableOffsetsFromProcessing) {
-        ensureStoresRegistered();
-
-        // write the checkpoint file before closing
-        if (checkpointFile == null) {
-            checkpointFile = new OffsetCheckpoint(new File(baseDir, CHECKPOINT_FILE_NAME));
-        }
-
-        updateCheckpointFileCache(checkpointableOffsetsFromProcessing);
+    public void checkpoint(final Map<TopicPartition, Long> writtenOffsets) {
+        // first update each state store's current offset, then checkpoint
```

Comment:
We keep the state store metadata's offset as the source of truth of the current position: during restoring this is updated from the last record offset from `restore`, during normal processing the `writtenOffsets` are passed in from record collector, and used to update the source of truth first. Then we just dump the source-of-truth map to the checkpoint file.